salt/modules/kubernetes.py

```diff
@@ -0,0 +1,500 @@
+'''
+Module for handling kubernetes calls.
+
+:optdepends:    - kubernetes Python client
+:configuration: This module is not usable until the following are specified
+    either in a pillar or in the minion's config file::
+
+        kubernetes.user: admin
+        kubernetes.password: verybadpass
+        kubernetes.api_url: 'http://127.0.0.1:8080'
+'''
+# Import Python Futures
+from __future__ import absolute_import
+from six import iteritems
+from salt.exceptions import CommandExecutionError
+import salt.utils
+import salt.utils.templates
+import yaml
+
+try:
+    import kubernetes
+    import kubernetes.client
+    from kubernetes.client.rest import ApiException
+
+    HAS_LIBS = True
+except ImportError:
+    HAS_LIBS = False
+
+
+import logging
+log = logging.getLogger(__name__)
+
+__virtualname__ = 'kubernetes'
+
+
+def __virtual__():
+    '''
+    Check dependencies
+    '''
+    if HAS_LIBS:
+        return __virtualname__
+
+    return False
+
+
+def ping():
+    '''
+    Checks connections with the kubernetes API server.
+    Returns True if the connection can be established, False otherwise.
+
+    CLI Example:
+        salt '*' kubernetes.ping
+    '''
+    status = True
+    try:
+        nodes()
+    except:
+        status = False
+
+    return status
+
+
+def nodes():
+    '''
+    Return the names of the nodes composing the kubernetes cluster
+
+    CLI Examples::
+
+        salt '*' kubernetes.nodes
+    '''
+    _setup_conn()
+
+    api_instance = kubernetes.client.CoreV1Api()
+    api_response = api_instance.list_node()
+
+    ret = []
+
+    for node in api_response.items:
+        ret.append(node.metadata.name)
+
+    return ret
+
+
+def _setup_conn():
+    '''
+    Setup kubernetes API connection singleton
+    '''
+    host = __salt__['config.option']('kubernetes.api_url',
+                                     'http://localhost:8080')
+    username = __salt__['config.option']('kubernetes.user')
+    password = __salt__['config.option']('kubernetes.password')
+    kubernetes.client.configuration.host = host
+    kubernetes.client.configuration.user = username
+    kubernetes.client.configuration.passwd = password
+
+
+def show_deployment(name, namespace):
+    '''
+    Return the kubernetes deployment defined by name and namespace
+
+    CLI Examples::
+
+        salt '*' kubernetes.show_deployment my-nginx,default
+        salt '*' kubernetes.show_deployment name=my-nginx,namespace=default
+    '''
+    _setup_conn()
+    try:
+        api_instance = kubernetes.client.ExtensionsV1beta1Api()
+        api_response = api_instance.read_namespaced_deployment(name, namespace)
+
+        return api_response.to_dict()
+    except ApiException as exc:
+        if exc.status == 404:
+            return None
+        else:
+            print(
```

Comment:
We prefer not to print from execution modules. Instead, either use `log.exception(exc)` and we'll format and log the exception for you.